### DISCUSSION TITLE
Zod validation input type

### DISCUSSION THREAD

**Discussion by Ranzeb** [colinhacks/zod#1995] [Q&A]
I'm new to zod and I'm looking for a way to have a specific input type and output type (the same) while my parse function can accept different types.

For example, having a zod schema that has z.input type equal to obj or undefined but I can also parse string as value and then make some transformation before type check in order to keep the input type as I want without considering the string type parsed.

Is it possible? Am I approaching the problem in the wrong way?

**Comment by JacobWeisenburger** [comment]
Are you able to give a code example?

**Reply by Ranzeb** [reply]
I've written this code to check if my input data return a specific type that can be either ObjectId or undefined.

`const stringObjSchema = z.object({
  _id: z.instanceof(ObjectId).or(z.string().transform((val) => { return ObjectId.createFromHexString(val) })).or(z.undefined())
})`

Now, this schema can accept both strings, ObjectId, and undefined then return the correct type.

When I check the inferred type with z.infer, z.input, and z.output I would like to have an emit that returns ObjectId or Undefined but of course, in the input type, I also have a string.

Is there a way to preprocess the string to have the same input and output type?

**Reply by JacobWeisenburger** [reply]
Are you able to give a complete code example that I can paste into my IDE and has all parts included so I can run it?

The example you sent is not including `ObjectId`
```ts
const stringObjSchema = z.object( {
    _id: z.instanceof( ObjectId ).or( z.string().transform( val => {
        //             ^^^^^^^^
        // Cannot find name 'ObjectId'.
        return ObjectId.createFromHexString( val )
        //     ^^^^^^^^
        // Cannot find name 'ObjectId'.
    } ) ).or( z.undefined() )
} )
```

**Reply by Ranzeb** [reply]
I'm sorry, here is my snippet code. All the type variables should be of type ObjectId | undefined

```ts
import { z, ZodTypeDef } from 'zod';
import { ObjectId } from 'bson';

const mySchema = z.object({
  //_id: z.instanceof(ObjectId).or(stringTypeCheck).or(z.undefined())
  _id: z.instanceof(ObjectId).or(z.undefined()).or(z.string().transform((val) => { return ObjectId.createFromHexString(val) }))
})

const hexString = "be5d55129807e725bda1b061";

type correctType = z.infer<typeof mySchema>
type correctInputType = z.input<typeof mySchema>
type correctOutputType = z.output<typeof mySchema>


const actualNodeDataFromObjectId = mySchema.parse({
  _id: ObjectId.createFromHexString(hexString)
})

const actualNodeDataFromString = mySchema.parse({
  _id: hexString
})
```

**Comment by JacobWeisenburger** [comment]
Everything seems to be working as expected. Can you explain what is not working as you are expecting?

```ts
import { z } from 'zod'
import { ObjectId } from 'bson'

const mySchema = z.object( {
    _id: z.instanceof( ObjectId )
        .or( z.undefined() )
        .or( z.string().transform( ObjectId.createFromHexString ) )
} )

type correctType = z.infer<typeof mySchema>
type correctInputType = z.input<typeof mySchema>
type correctOutputType = z.output<typeof mySchema>

const hexString = 'be5d55129807e725bda1b061'

const actualNodeDataFromObjectId = mySchema.parse( {
    _id: ObjectId.createFromHexString( hexString )
} )
console.log( actualNodeDataFromObjectId._id?.toHexString() )
// be5d55129807e725bda1b061

const actualNodeDataFromString = mySchema.parse( {
    _id: hexString
} )
console.log( actualNodeDataFromString._id?.toHexString() )
// be5d55129807e725bda1b061
```

**Reply by Ranzeb** [reply]
If you check the correctInputType emit something like ObjectId | string | undefined. I need to have it be like the correctType or correctOutputType and be like ObjectId | undefined. 

Basically, I'm looking for a way to accept string and preprocess them to be converted into ObjectId in order to have my InputType emit only ObjectId | undefined. 

Hope this is much clearer

**Reply by JacobWeisenburger** [reply]
The point of the input type is to tell you what is an acceptable type to pass into parse. So removing string from the input type but yet allowing it to receive strings seems like a bad practice. Why do you want your input type not to match the things that can be used as inputs?

**Reply by Ranzeb** [reply]
I'm trying to understand because I'm new to Zod, please sorry if it seems like bad practice, and seems like my code works right so probably I already have a good answer.

By the way, my idea was to be sure to check only if the input is an ObjectId or not. I don't care about strings if they aren't converted before to ObjectId. Something like a checker if ObjectId is valid or not.

**Comment by JacobWeisenburger** [comment]
Thanks for your patience while I work to understand what you are wanting. I'm pretty sure you want something like this. It's ok that `Input` accepts a string, because it will be transformed to an ObjectId.

```ts
import { z } from 'zod'
import { ObjectId } from 'bson'

const objectIDSchema = z.instanceof( ObjectId ).optional().or(
    z.string().transform( ( input, ctx ) => {
        try {
            return ObjectId.createFromHexString( input )
        } catch ( error ) {
            ctx.addIssue( {
                message: error.message,
                code: z.ZodIssueCode.custom,
                params: { input },
            } )
        }
    } )
)

type Input = z.input<typeof objectIDSchema>
// type Input = string | ObjectId | undefined

type Output = z.output<typeof objectIDSchema>
// type Output = ObjectId | undefined

const hexString = 'be5d55129807e725bda1b061'
const objectID = ObjectId.createFromHexString( hexString )

console.log( objectIDSchema.safeParse( hexString ).success ) // true
console.log( objectIDSchema.safeParse( objectID ).success ) // true

const invalidResult = objectIDSchema.safeParse( 'foo' )
!invalidResult.success && console.log( invalidResult.error.issues )
// [
//     {
//         message: 'Argument passed in must be a single String of 12 bytes or a string of 24 hex characters',
//         code: 'custom',
//         params: { input: 'foo' },
//         path: []
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Ranzeb** [reply]
What's the role of ctx in this code? It represent the context of the application?

**Reply by JacobWeisenburger** [reply]
It's just the context of the parse. So all it allows you to do is add issues to the final errors obj and check the current property path.
More info here: https://github.com/colinhacks/zod#validating-during-transform

**Reply by Ranzeb** [reply]
Thank you Jacob, I will keep investigate on your link

**Reply by JacobWeisenburger** [reply]
You're welcome and welcome to the Zod community.

**Comment by Ranzeb** [comment]
Hi, @JacobWeisenburger I've still worked on this topic and found a working solution.

```ts
const PreprocessType = z.preprocess((arg) => {
  if (typeof arg == "string") return ObjectId.createFromHexString(arg)
  else return arg;
}, z.instanceof(ObjectId).optional())

type Input = z.input<typeof PreprocessType> & (ObjectId | undefined);
type infer = z.infer<typeof PreprocessType> & (ObjectId | undefined);
type Output = z.output<typeof PreprocessType> & (ObjectId | undefined);

const FinalDataType: z.ZodType<Output, z.ZodTypeDef ,Input> = z.object({
  _id: PreprocessType
})

type correctPreprocessTypeInput = z.input<typeof FinalDataType>
type correctPreprocessType = z.infer<typeof FinalDataType>
type correctPreprocessTypeOutput = z.output<typeof FinalDataType>

const actualNodeDataFromObjectId = FinalDataType.parse({
  _id: ObjectId.createFromHexString(hexString)
})

const actualNodeDataFromString = FinalDataType.parse({
  _id: hexString
})

```
The Input and Output types are correct (ObjectId or Undefined), and the .parse() work fine.

If you try this code on your IDE you can see that there's a typescript error on FinalDataType referring to incompatible types. Do you have any idea how to fix this typescript error? Because everything now works fine and solve my problem perfectly

**Reply by Ranzeb** [reply]
I've found the solution! 

I will probably post it tomorrow

**Reply by Ranzeb** [reply]
This is my solution, however, I'm implicitly casting the output type. 

Is there a way to achieve the same results without implicitly casting the output?

```ts

import { z } from 'zod';
import { ObjectId } from 'bson';
//import NodeDataType from './schema/nodeDataType';

const hexString = "be5d55129807e725bda1b061";


const PreProcessId =
  z.preprocess((arg) => {
    if (typeof arg == "string" && arg.length >= 24) return ObjectId.createFromHexString(arg)
    else if (arg instanceof ObjectId) return arg;
    else return undefined;
  }, z.instanceof(ObjectId).optional()) as z.ZodEffects<z.ZodTypeAny, ObjectId | undefined, ObjectId | undefined>;


const NodeDataType = z.object({
  _id: PreProcessId
});

type correctType = z.infer<typeof NodeDataType>
type correctInputType = z.input<typeof NodeDataType>
type correctOutputType = z.output<typeof NodeDataType>

const actualNodeDataFromObjectId = NodeDataType.parse({
  _id: ObjectId.createFromHexString(hexString)
})

const actualNodeDataFromString = NodeDataType.parse({
  _id: hexString
})

console.log(actualNodeDataFromObjectId);
console.log(actualNodeDataFromString);
```

**Reply by Ranzeb** [reply]
@JacobWeisenburger Do you have any idea? I've found another solution using z.custom<>() but not sure if it a correct ones
